More refactor and unit test: Server & Client refactoring

Manghi opened the pull request:
I finished with the major refactoring for Server.rs. 
Most of the changes here are for breaking up the chat message portions of  an `Update`.

Some unit tests were added too. I started prepending each unit test with the function under test, so the full name ends up as `function_name_concept_we_are_testing()`. It's a bit easier to track since some of our tests have quite a lot of setup.

Since this contains a considerable amount of code, I figure we should push this in before I move on. I think there are other unit tests that I would like to add for Server but I need to look when I am less tired.

Manghi committed:
Refactored chat messaging into components

Manghi committed:
Added unit tests for `get_message_skip_count()`

Manghi committed:
Man that comment made no sense once I pushed it in; Fixed

Manghi committed:
Refactored and unit tests: collect_unacknowledged_messages()

Manghi committed:
Tests for handle_chat_message()

Manghi committed:
create_new_room() tests

Manghi committed:
join_room() and leave_room() tested

Manghi committed:
add_new_player(), is_unique_player_name(), expire_old.. unit tests

AaronM04 commented on src/server.rs:
```diff
@@ -905,5 +984,550 @@ mod test {
+    }
+/*
+    #[test]
+    fn expire_old_messages_in_all_rooms_old_messages_are_wiped()
```
FYI, if a test compiles but does not pass yet, you can add `#[ignore]` to the front. That's not the case here since it doesn't compile.

Manghi replied:
Cool good to know.

AaronM04 commented on src/server.rs:
```diff
@@ -578,78 +635,89 @@ impl ServerState {
     }
 
+    /// Creates a vector of messages that the provided Player has not yet acknowledged.
+    /// Exists early if the player is already caught up.
```
please fix type: "Exists" -> "Exits"

AaronM04 commented on src/server.rs:
```diff
@@ -220,14 +224,67 @@ impl Room {
+    }
+
+    /// This function retrieves the number of messages FIFO which has
+    /// already been acknowledged by the client.
```
This comment is a little confusing. What is "number of messages FIFO"?

Manghi replied:
FIFO shouldn't be there. Will be fixed and reworded a bit in latest commit.

AaronM04 commented on src/server.rs:
```diff
@@ -220,14 +224,67 @@ impl Room {
     }
+
+    /// Gets the oldest message in the room message queue
+    fn get_oldest_msg(&self) -> Option<&ServerChatMessage> {
```
You can replace the body of this function with just `self.messages.front()`, since that returns an Option.

AaronM04 commented on src/server.rs:
```diff
@@ -220,14 +224,67 @@ impl Room {
+    }
+
+    /// Gets the newest message in the room message queue
+    fn get_newest_msg(&self) -> Option<&ServerChatMessage> {
```
Same as above.

AaronM04 commented on src/server.rs:
```diff
@@ -578,78 +635,89 @@ impl ServerState {
+                    let amount_to_consume = room.get_message_skip_count(chat_msg_seq_num);
+
+                    // Cast to usize is safe because our message containers are limited by MAX_NUM_CHAT_MESSAGES
+                    let mut message_iter = room.messages.iter();
```
This doesn't need to be `mut`. Also I would suggest not even creating `messages_iter` and combine this line with the next, but it's your call. :)

Manghi committed:
expire_old_messages_in_all_rooms expire test cases

Manghi committed:
handle_new_connection unit tests

Manghi committed:
handle_new_connection unit tests

AaronM04 reviewed (changes requested):
Amazing work! I had some feedback. Other than the feedback, everything so far looks good.

Manghi committed:
Addressed valuable PR comments :)

Manghi committed:
Using proptest! for process_request_action()

Manghi committed:
Disabling this test until I determine why it panics

Manghi committed:
process_request_action() and handle_request_action() unit tests

Manghi committed:
construct_client_updates() unit tests

Manghi committed:
Updated versioning to use semver, fixed unit tests!

Manghi committed:
Silencing warning during cargo build

Manghi commented:
@AaronM04  Thanks for the great review and positive feedback. I addressed your comments in 540f261. 

I also added a few more unit tests. I am officially satisfied with what we have for `server.rs` and will move onto `client.rs` shortly.

Please review when you get a chance.

Manghi committed:
Fine you get a new line...

Manghi committed:
Use net.rs::bind in client

Manghi committed:
ResponseCode::OK case refactored and tested

Manghi committed:
RequestAction::LoggedIn refactored and tested

Manghi committed:
Refactored player and room list

Manghi committed:
Packet::Update chat handler moved into own function

Manghi committed:
Refactored command translation into request action

- Added a QuitGame request action which will only be used on client side

Manghi committed:
Refactored Event::Response into own function

Manghi committed:
Refactored TickEvent and UserInputEvent

Manghi commented:
Scratch that last comment... I started working on `client.rs` because it wasn't that much work. All that remains is to unit test some of these new functions and then I'll be completely done.

Manghi committed:
handle_incoming_chats() unit tests

Manghi committed:
parse_stdin_input() unit tests

Manghi committed:
build_command_request() unit tests

Manghi committed:
handle_user_input_event() unit test

Manghi commented:
@AaronM04 
I am happy with the netwayste server and client refactoring/unit tests. Whew! 

Please review when you get a chance. Danke.

AaronM04 commented on src/net.rs:
```diff
@@ -45,6 +48,7 @@ pub enum RequestAction {
     NewRoom(String),
     JoinRoom(String),
     LeaveRoom,
+    QuitGame,
```
Isn't this already covered by `LeaveRoom`?

Manghi replied:
`LeaveRoom` keeps the player connected and in lobby. `QuitGame` would tell the Server that this Client has decided to gracefully terminate the connection.

AaronM04 commented:
Yeah, but we have `Disconnect` for that.

On July 4, 2018 10:14:34 PM PDT, Manghi <notifications@github.com> wrote:
>Manghi commented on this pull request.
>
>
>
>> @@ -45,6 +48,7 @@ pub enum RequestAction {
>     NewRoom(String),
>     JoinRoom(String),
>     LeaveRoom,
>+    QuitGame,
>
>`LeaveRoom` keeps the player connected and in lobby. `QuitGame` would
>tell the Server that this Client has decided to gracefully terminate
>the connection. 
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/conwayste/netwayste/pull/4#discussion_r200235261

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

Manghi committed:
No need for QuitGame, use existing Disconnect!!!

AaronM04 reviewed (approved):
LGTM
